\documentclass{article}
\usepackage{graphicx} % Required for inserting images
\usepackage[a4paper,top=2cm,bottom=2cm,left=3cm,right=3cm,marginparwidth=1.75cm]{geometry}
\setlength {\marginparwidth }{2cm} % remove warning about todonotes
\usepackage{todonotes} %Enables ToDo notes (\todo)
%\usepackage{minted}
\usepackage{placeins}
\usepackage[authoryear]{natbib}
\usepackage{caption} % to adjust caption width
\usepackage{amsmath}

\bibliographystyle{myapalike}

\usepackage[hidelinks]{hyperref} % links in the text, should be loaded last


% Tile Page Setup
\title{BINP28 Project Report - Genetic relationships from a PCA}
\author{Max Weber}
\date{Feb 2026}

\begin{document}

% Title Page
\maketitle
\tableofcontents
\newpage

\section{Methods}

The analysis is based on a .vcf file with 16 samples (Chromosome 5 and Z) from 4 population, of which one population is an outgroup with only one sample.
Variant calling was already performed before this analysis, no detail on the upstream workflow or the origin of the samples are known.
Before running a PCA on the variants, filtering of the variants was done with bcftools v. 1.23 \citep{bcftools, samtools}.
The outgroup sample was removed and each sample and site combination was filtered based on sequencing depth, data with less then 3 or more then 14 reads was excluded.

The sequencing depth threshold was based on the distribution of reads.
Bcftools automatically adjusts the AN (count of sites with known alleles) AC (count of alternative alleles) values based on the masking, make subsequent filtering based on AN = 30 and AC $>$ 2 possible. The AN filtering excludes all sites for which not all samples have data or samples that are masked from the previous step, as we have 15 diploid samples, yielding AN = 30 only if all samples have available available data. The filtering reduces the number variants from 3,816,977 to 112,329 variants.

Further analysis of the filtered data was performed in R v 4.5.2 \citep{RCore} using the package ``tidyverse'' for data handling and plotting \citep{tidyverse} and the package ``SNPRelate'' for LD (linkage disequilibrium) pruning and PCA \citep{SNPRelate}.
SNPRelate was chosen due to fast performance compared to alternatives due to multi-threading support and it's ability to handle different variant types \citep{SNPRelate}.

The filtered variants were converted from vcf to SNPRelate's gds format, LD pruning was performed on the whole genome (both available chromosomes) with a threshold of r = 0.2. This excluded a large proportion of the samples reducing the number of variants from 112,329 to just 2,354. This number is suspiciously low, and trying

  vdsvÂ´

\FloatBarrier
\section{Results}


\begin{figure}[h]
    \centering
    \includegraphics[width=0.9\linewidth]{../04_pca/pc12.png}
    \caption{PC1 and PC2 of PCA run on genetic variances among 15 individuals from three different populations. The true population names are unknown, instead, the common start of the sample names are used.}
    \label{pc12}
\end{figure}


\begin{figure}[h]
    \centering
    \includegraphics[width=0.85\linewidth]{../04_pca/elbow_plot.png}
    \caption{Scree plot of PCA run on genetic variances among 15 individuals from three different populations. The x-axis gives the number of the PC, the y-axis shows how much variance is explained by a PC in \%.}
    \label{scree_plot}
\end{figure}

\FloatBarrier
\section{Discussion}

- Variance homogenous for all PCs => does fit expected ``elbow shape'' (but does fit at the start with PC1, so overall it fits, just not a sharp elbow)
- Last PC (15) artifact, because we have 15 samples, we should only have 14 PC, did not find option to limit PCs in SNPRelate

% References
\newpage
\bibliography{references}

\end{document}

